Serialize and deserialize trajectory builder options

ojura opened the pull request:
The options proto is written as a second proto in the stream, behind the pose graph proto.

SirVer commented:
@ojura Is this something you want to get in? It seems to me that the TrajectoryBuilderOptions are configuration, expressed in Lua and not State. It seems peculiar to me to serialize it as state.

SirVer commented:
@cschuet @wohe I have no strong feelings either way about this. What do you think?

wohe commented:
It seems this reverts the fix to #249. Why do we deserialize the options that were once used but no longer are? @ojura I also do not understand why your asset writer wants to reuse the resolution that was used for SLAM. Is this not a separate choice, i.e. why do you not add this as an option to your asset writer?

ojura commented:
Sure, I can handle the resolution option externally, but that is not the only thing from here that is used: for example, also `RangeDataInserterOptions`.

The submaps that are deserialized into libcartographer were previously built with certain (obviously immutable) `LocalTrajectoryBuilderOptions` (including `RangeDataInserterOptions`). After deserialization, the built submaps are restored, but you don't know which `LocalTrajectoryBuilderOptions` were used for building them, should you need it to inspect it for some reason. (And I do, because I want to reuse all these options for rebuilding the map.)

I can of course handle all of this externally, of course. That's what I'm going to do if the PR doesn't go through :-)

> It seems this reverts the fix to #249.

Yes, because since we have the trajectory builder options data available in the proto stream, and restoring it is very cheap, I don't see a reason not to restore these options.

ojura commented:
I think I came up with a good analogy: when a camera takes a picture, it records the focal length, aperture size, exposure and ISO. You can't really change these parameters, because this would mean taking another picture. But you _can_ inspect these parameters and make some conclusions on how they influence the quality of the picture.

Cartographer recording the options used for building the trajectory and the submaps would be analogous.

cschuet commented:
@ojura I have looked at this again and I must say that I am really deeply uncomfortable with deserializing the `TrajectoryBuilderOptions`. I completely get your analogy and I agree that we should add the options when writing the pbstream, so that your pbstream management GUI can display them or whatever you want to do with them, but I wouldn't want to deserialize them in LoadMap.

ojura commented:
Okay, sure. I will update this to avoid full deserialization.

ojura commented:
@cschuet I reworked this a little, can you take a look? You didn't like full deserialization of a new trajectory builder using the saved options, so I removed it. I also reverted removing `AddTrajectoryForDeserialization`, and expanded it to just pass through the already previously serialized options.

ojura commented on cartographer/mapping/map_builder.cc:
```diff
@@ -227,12 +251,20 @@ void MapBuilder::SerializeState(io::ProtoStreamWriter* const writer) {
-  proto::PoseGraph pose_graph;
-  CHECK(reader->ReadProto(&pose_graph));
+  proto::PoseGraph pose_graph_proto;
+  CHECK(reader->ReadProto(&pose_graph_proto));
```
I think it would make sense to introduce a pbstream serialization header proto, which would include the pose graph and trajectory builder options protos.

ojura commented:
@cschuet updated. This is blocking #549 for almost two months, can we resolve this?

ojura commented:
Shoot, I messed up by not overriding the interface method. Will fix and reopen.
